### PR TITLE
7903449: Jextract generates structs that cannot be compiled

### PR DESCRIPTION
**Pull request by mcimadamore**
While most of the constants required by jextract are stored in the auxiliary constant$xyz files, some constants are stored inside the class declaration corresponding to a struct. This was done to minimize the number of constant classes being generated. However, this can backfire, as when the generated struct is big, too many constants are added to it, which can result in a source code which fails to compile (as it's too big).

Instead of doing something even more convoluted, I've decided to centralize the generation of _all_ constants under the constant$xyz files. This means that there should be no constants generated in structs (or any other "public" declaration).

Doing this was problematic, as the current arrangement somehow limits the name clashes when adding constants with same name to different structs. After some attempts, I realized that the best way forward would be to simplify the `ConstantBuilder` class, so that it no longer cares about "public" Java names. Instead that class should only be used to generate declaration for constants that are used internally, by the jextract bindings. Because of that, _any_ name will do - for this reason this patch switches to a regular naming schemes for all constants, like `const$0`, `const$1`, etc.

Note that these are **not** the names exposed by jextract bindings - such names (the names of the constant *getters*) are still the same as before.

The main changes are around `ConstantBuilder`, which is now gone. In its place we find `Constants`, which acts as a sort of source-based constant pool. `Constants` has a cache (where already defined constants of certain kinds are stored), and new constant declarations are only emitted when needed. I've also simplified the API which is used by clients to generate constants, to remove the need of lambda expressions (which seemed a bit over the top).

I've also centralize the generation of numeric constants under `Constants` as well (before, these constants where generated on the fly in `HeaderFileBuilder`).

And, I also realized that our setup for sharing primitive layout constants was too complex. Most primitive layout constants are just wired to some constant in `ValueLayout`. The only special case is `C_POINTER`, which needs its *target layout* set (and presumably we only want to do that once). A much simpler solution is to just store `C_POINTER` inside `RuntimeHelper` and just use that.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (no review required)

### Issue
 * [CODETOOLS-7903449](https://bugs.openjdk.org/browse/CODETOOLS-7903449): Jextract generates structs that cannot be compiled

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jextract.git pull/117/head:pull/117` \
`$ git checkout pull/117`

Update a local copy of the PR: \
`$ git checkout pull/117` \
`$ git pull https://git.openjdk.org/jextract.git pull/117/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 117`

View PR using the GUI difftool: \
`$ git pr show -t 117`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jextract/pull/117.diff">https://git.openjdk.org/jextract/pull/117.diff</a>

</details>

### CODE REVIEW

**Comment by mcimadamore** [comment]
I've uploaded a new revision which improves how primitive constants (e.g. numerics) are handled. Now constants can be either "named" (e.g. a field in a class) or "immediate" (e.g. just an expression on the fly). This way, we can model certain constants in a more direct form, w/o the need of creating a new field. Generating fields for all basic primitive constants balooned up the number of constant classes in an unacceptable way.

**Review by JornVernee — COMMENTED** [review]
Very nice

File: src/main/java/org/openjdk/jextract/impl/Constants.java [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+    }
+
+    public Constant addDowncallMethodHandle(String nativeName, FunctionDescriptor descriptor, boolean isVarargs) {
+        return builder().emitDowncallMethodHandleField(nativeName, descriptor, isVarargs, false);
+    }
+
+    public Constant addVirtualDowncallMethodHandle(FunctionDescriptor descriptor) {
+        record DowncallKey(FunctionDescriptor desc) { }
+        DowncallKey downcallKey = new DowncallKey(descriptor);
+        Constant constant = cache.get(downcallKey);
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/main/java/org/openjdk/jextract/impl/Constants.java:506]: Likely, if we want to deduplicate fully. I'm a bit worried about doing a full traversal and creating a brand new function descriptor just for the lookup though. I guess I could implement some custom equals method in the key.

[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/main/java/org/openjdk/jextract/impl/Constants.java:506]: After some more consideration, I decided to leave this as is. Reason being that the benefit would be somewhat limited - note that this method is only used for functional upcall interfaces. So, for this to make a difference you'd need to have several function pointers taking different structs by value with the same underlying layout (modulo names) - which seems a bit obscure. If needed we can fix later.

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**